frontend/model/contracts/group.js

```diff
@@ -312,32 +312,138 @@ sbp('chelonia/defineContract', {
       }
     },
+    // May return 'undefined', in which case the caller should check archived data.
     periodBeforePeriod (state, getters) {
       return (periodStamp: string) => {
-        const len = getters.groupSettings.distributionPeriodLength
-        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), -len))
+        if (!isPeriodStamp(periodStamp)) throw new TypeError('must be periodStamp')
+        const { distributionDate, distributionPeriodLength } = getters.groupSettings
+        if (!distributionDate) return
+        // This is not always the current period stamp.
+        const distributionDateStamp = dateToPeriodStamp(distributionDate)
+        const sortedPeriodKeys = getters.groupSortedPeriodKeys
+        const latestKnownStamp = sortedPeriodKeys[sortedPeriodKeys.length - 1]
+        // Maybe the distribution date has just been updated,
+        // but the corresponding period has not started or is not stored yet.
+        // So we try to use the distribution date first.
+        if (periodStamp > distributionDateStamp) {
+          // Only extrapolate one period length in the future.
+          const onePeriodLenghtAhead = addTimeToDate(distributionDate, distributionPeriodLength)
+          if (periodStamp === dateToPeriodStamp(onePeriodLenghtAhead)) return distributionDate
+          else return
+        }
+        if (periodStamp === distributionDateStamp) {
+          if (sortedPeriodKeys.length) {
+            // If the distribution date doesn't match the latest known period stamp,
+            // then either that stamp is for the waiting period,
+            // or the distribution date has just been updated.
+            // In both cases we can return it.
+            if (latestKnownStamp !== distributionDateStamp) return latestKnownStamp
+            // Otherwise it's a normal period, therefore substracting the period length would not be reliable.
+            else return sortedPeriodKeys[sortedPeriodKeys.length - 2] ?? undefined
+          } else {
+            // If no period has been stored yet, then we're in the waiting period and can do arithmetic.
+            return dateToPeriodStamp(addTimeToDate(distributionDate, -distributionPeriodLength))
+          }
+        }
+        const index = sortedPeriodKeys.indexOf(periodStamp)
+        if (index === -1) {
+          // Maybe the given stamp is wrong and has no associated period,
+          // but is one period length ahead of a known one.
+          // TODO: just return 'undefined' when sure it's always safe.
+          const maybePreviousStamp = dateToPeriodStamp(
+            addTimeToDate(dateFromPeriodStamp(periodStamp), -distributionPeriodLength)
+          )
+          return sortedPeriodKeys.includes(maybePreviousStamp) ? maybePreviousStamp : undefined
+        }
+        // If index is 0 then the caller will have to check the archive.
+        if (index === 0) return
+        return sortedPeriodKeys[index - 1]
```

Comment:
I do not understand what is going on here or why this code is so complicated.

AFAICT this is what this function should be:

```js
    periodBeforePeriod (state, getters) {
      return (periodStamp: string) => {
        const { distributionPeriodLength } = getters.groupSettings
        const sortedPeriods = getters.groupSortedPeriodKeys
        for (const i = sortedPeriods.length - 1; i >= 0; --i) {
          const latestPeriod = sortedPeriods[i]
          if (periodStamp >= latestPeriod) return latestPeriod
        }
        return dateToPeriodStamp(addTimeToDate(dateFromPeriodStamp(periodStamp), -distributionPeriodLength))
      }
    }
```

Reply:
If I'm not mistaken, this code won't ever return `undefined` to defer handling to the archive

Reply:
OK, that's a good point regarding `distributionDate` - my code doesn't check for that because it assumes (incorrectly) that that is already stored in `sortedPeriods`, which you're absolutely right, it might not be.

Also, my code has a bug related to this check:

```js
if (periodStamp >= latestPeriod) return latestPeriod
```

I think that should be:

```js
if (periodStamp > latestPeriod) return latestPeriod
```

The code that's currently here in the PR needs to be changed for a few reasons:

1. As discussed above, we need to avoid passing `distributionDate` to `dateToPeriodStamp`
2. It needs to be simplified similarly to what I wrote above. Right now it's got some very questionable lines, like this:

```js
        if (periodStamp > distributionDateStamp) {
          // Only extrapolate one period length in the future.
          const onePeriodLenghtAhead = addTimeToDate(distributionDate, distributionPeriodLength)
          if (periodStamp === dateToPeriodStamp(onePeriodLenghtAhead)) return distributionDate
          else return
        }
```

This is very strange. The purpose of this function is to return the period before a given period stamp. That's not what the lines above do at all. Instead they seemingly randomly add a period length to `distributionDate` and compare `periodStamp` to that, and then bails if the check fails.

We need something conceptually equivalent to the code I wrote above that:

1. Takes into account `distributionDate`
2. Has a consistent mental model for how period stamps are created and iterated forward/backward

## Mental model

Our mental model is as follows:

```
<--- [ UNKNOWN PAST ] --- [ SORTED_KEYS ] ---- [ UNKOWN FUTURE ] --->
```

Here we see 3 section. There is a "known" timeline in the center, and an unknown timeline at each end.

If we are given a period stamp that is inside of `[ SORTED_KEYS ]`, here's what iterating it looks like:

```
[ A ] <-> [ B ] <-> [ C ]
                 ^
                 |
         periodStamp
```

If we want to go forward in time from `periodStamp` pointing at `[ B ]` stamp, then `periodAfterPeriod` points us to `[ C ]` and `periodBeforePeriod` points us to `[ A ]`.

If, however, we have this situation:

```
[ B ] <-> [ C ] <-> [ UNKNOWN FUTURE ]
                 ^
                 |
         periodStamp
```

Then:

- `periodAfterPeriod` adds `distributionPeriodLength` to `periodStamp` and returns the result
- `periodBeforePeriod` returns `[ B ]`

And, if we have this situation:

```
[ UNKNOWN PAST ] <-> [ A ] <-> [ B ]
                        ^
                        |
                   periodStamp
```

Then:

- `periodAfterPeriod` returns `[ B ]`
- `periodBeforePeriod` returns `undefined`

Note: that this is the **only scenario** where `undefined` can be returned.

Also note: in that case `historicalPeriodBeforePeriod` (please rename from currently called `getPeriodBeforePeriod`) **does the exact same logic as `periodBeforePeriod`**, except it performs the check on a version of `[ SORTED_KEYS ]` that includes historical payments from then archive.

In fact, because both `periodBeforePeriod` and `historicalPeriodBeforePeriod` contain line-by-line exactly the same logic, please create a shared function inside of `time.js` that does this logic. When `historicalPeriodBeforePeriod` and `periodBeforePeriod` call that function, they will include an array containing the `[ SORTED_KEYS ]`.

Finally, to take into account the `distributionDate`, simply add in the `distributionDate` to the `[ SORTED_KEYS ]` that you pass in to the internal `periodBeforePeriod` inside of `time.js`.

And similarly, there will be a corresponding shared `periodAfterPeriod` in `time.js` that is called from both `group.js` and `PaymentsMixin.js`.

Hope that's clear! Let me know if you have any questions.

Reply:
> This is very strange. The purpose of this function is to return the period before a given period stamp.

Exactly, so if the given date is not actually a past, current or future period stamp but just a random ISO string, my assumption was that `undefined` had to be returned.

If that assumption shoud be dropped so that these getters behave like `periodStampGivenDate`, e.g. `periodAfterPeriod` returns `B` for any date between `A` and `B` not just for `A`, then maybe they could as well be renamed for consistency:
- nextPeriodStampGivenDate
- periodStampGivenDate
- previousPeriodStampGivenDate
(I'd rather use -forDate than -givenDate). Otherwise `periodAfter/BeforeDate` also sound good to me.

Another assumption in this PR, is that `distributionPeriodLength` is now a variable setting whose value only holds for the current period. It can no longer be safely used to iterate between existing periods, but is as a hint to initially setup a defof subtracted from past periods.
Moreover, in the waiting period the distribution date i.e. the next period stamp can be updated without updating `distributionPeriodLength` accordingly, because s really just a hint, to set a default next distribution date when entering a new period.